### PR TITLE
fix: validate digest prior to layer download

### PR DESCRIPTION
**Pull request by RTann**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -183,7 +183,7 @@ require (
 replace (
 	github.com/facebookincubator/nvdtools => github.com/stackrox/nvdtools v0.0.0-20231111002313-57e262e4797e
 
-	github.com/heroku/docker-registry-client => github.com/stackrox/docker-registry-client v0.1.0
+	github.com/heroku/docker-registry-client => github.com/stackrox/docker-registry-client v0.2.1-0.20240809225852-a122013c5989
```
Comment (RTann) [inline go.mod:186]: https://github.com/stackrox/docker-registry-client/pull/21 should be merged and v0.2.1 should be cut prior to merging this PR

**Review by dcaravel — APPROVED** [review]
LGTM - assuming tests pass and the dep updated to a tagged version as mentioned.

**Comment by RTann** [comment]
@dcaravel I updated the PR to add some more validation, as I fear the original was not sufficient. I believe now I cover all bases (fetching both manifests and layers. before was just layers).

File: pkg/scan/scan.go [hunk 2]
```diff
@@ -78,36 +81,36 @@ func ProcessImage(storage database.Datastore, image *types.Image, registry, user
...
+	return dig, storage.AddImage(layer, image.SHA, lineage, image.TaggedName(), &database.DatastoreOptions{
 		UncertifiedRHEL: uncertifiedRHEL,
 	})
 }
 
 // process fetches and analyzes the layers for the requested image returning the image digest, the lineage of the last layer and the last layer digest
 func process(storage database.Datastore, image *types.Image, reg types.Registry, uncertifiedRHEL bool) (string, string, string, error) {
 	logrus.Debugf("Processing image %s", image)
-	digest, layers, err := FetchLayers(reg, image)
+	dig, layers, err := FetchLayers(reg, image)
```
Comment (RTann) [inline pkg/scan/scan.go:99]: For anyone wondering, `image` definitely has a valid digest. See https://github.com/stackrox/scanner/blob/2.34.1/pkg/clairify/types/types.go#L178. This is called prior to both calls to `process`. This function returns an image which may have a digest. If it does, it is validated prior to returning it via `reference.ParseAnyReference`.

File: pkg/scan/scan.go [hunk 3]
```diff
@@ -44,6 +46,7 @@ func analyzeLayers(storage database.Datastore, registry types.Registry, image *t
 	for _, layer := range layers {
 		layerReadCloser := &LayerDownloadReadCloser{
 			Downloader: func() (io.ReadCloser, error) {
+				// It is assumed layer is a valid digest.Digest, so no need to parse and validate here.
```
Comment (jvdm) [inline pkg/scan/scan.go:49]: I am curious: Why you did not validate anyway to be on the safe side?

File: pkg/scan/scan.go (hunk 2 above)
Comment (jvdm) [reply pkg/scan/scan.go:99]: https://github.com/stackrox/scanner/blob/2.34.1/pkg/clairify/types/types.go#L181

File: pkg/scan/scan.go (hunk 3 above)
Comment (jvdm) [reply pkg/scan/scan.go:49]: Same for `handleManifest`.

**Review by jvdm — APPROVED** [review]
LGTM, just curious about why you expected validation outside the function calls? Not a blocker. I just think when this is the case, functions/methods/routines should still protect themselves from callers, but they don't need to support the invalid input by returning an error -- since this is a programmer's kinda error, they could just enforce and panic (like some standard library methods do for invalid input that cannot be handled in runtime).

File: pkg/scan/scan.go (hunk 3 above)
Comment (RTann) [reply pkg/scan/scan.go:49]: I originally validated here, but I removed it once I (believed) it was accounted for elsewhere. But you're right. Best to be safe

File: pkg/scan/scan.go [hunk 4]
```diff
@@ -78,36 +81,36 @@ func ProcessImage(storage database.Datastore, image *types.Image, registry, user
...
 }
 
 // process fetches and analyzes the layers for the requested image returning the image digest, the lineage of the last layer and the last layer digest
 func process(storage database.Datastore, image *types.Image, reg types.Registry, uncertifiedRHEL bool) (string, string, string, error) {
 	logrus.Debugf("Processing image %s", image)
-	digest, layers, err := FetchLayers(reg, image)
+	dig, layers, err := FetchLayers(reg, image)
 	if err != nil {
-		return digest, "", "", err
+		return dig, "", "", errors.Errorf("fetching layers for %s: %v", image.String(), err)
```
Comment (RTann) [inline pkg/scan/scan.go:101]: I forgot `errors.Wrapf` was a thing, so I'll use that in a followup commit

**Review by dcaravel — APPROVED** [review]

File: pkg/scan/scan.go (hunk 3 above)
Comment (RTann) [reply pkg/scan/scan.go]: For `handleManifest`, `ref` may not be a digest, but rather a tag, so I didn't want to do the check there

**Pull request merged**